### ISSUE TITLE
Rounding errors in the positional encodings

### ISSUE THREAD

**Issue by johahi** [lucidrains/enformer-pytorch#31]
Hi, many thanks for your efforts. 

I investigated the rounding errors a bit and to me it seems like they mostly accumulate in the gamma positional encodings. With same inputs (rate, concentration, x are torch.allclose), I get different results after the xlogy for tensorflow and pytorch, and because these are rather large deviations/values, this then seems to propagate. Unsure where TF and pytorch differ for xlogy here...

So as a "dirty" fix, I precomputed the gammas in TF, saved and loaded them in your pytorch implementation and I manage to get the same results (up to some numerical issues) with enformer-pytorch as if I compute it for some test sequences with the TF reference implementation. 

For instance, for this  [test_sequence.npy.zip](https://github.com/lucidrains/enformer-pytorch/files/12739399/test_sequence.npy.zip): the maximum of the output of the current pytorch implementation is 474.7, with the precomputed gammas 555.6, and the reference implementation has 555.5.

Feel free to have a look at my fork for all of this, where I integrated the precomputed gammas in your implementation.

**Comment by lucidrains** [comment]
wow, thank you for spending all this effort tracking down where the error was! could I possibly get a copy of your revised model so I can upload it to huggingface?

**Comment by lucidrains** [comment]
@johahi i decided to compute the gamma positions in float64. that reduced the error between the xlogy for torch and tensorflow from ~1e-6 down to 1e-14. if that doesn't help, we can resort to your solution!

**Comment by johahi** [comment]
@lucidrains thanks for the quick reponse, cool! 
I just ran the numbers for the example sequence above using your newest version, and unfortunately this does not seem to fix it. With float64, I get 454.9 as max, just to be comparable to what I report above (474.7). You can see however, how the errors accrue just from this little change. I actually suspect there is something more fundamentally incorrect about xlogy, but have not (yet?) looked further into it. 

I am a bit unsure about how to best integrate the precomputed gammas, maybe in the state dict, probably much nicer than explicitly loading them from disk - what do you think? I had to rewrite the positional encs a [bit](https://github.com/johahi/enformer-pytorch/blob/07f0e476a09b37a4e2f03d4046cbb175619b6699/enformer_pytorch/modeling_enformer.py#L72).

I reimplemented the enformer "successor" (borzoi) [in pytorch](https://github.com/johahi/borzoi-pytorch) which only uses central mask pos encodings. When I ported the weights there, things kinda just worked (it produces the same results as the TF model) and I thought why not revisit enformer-pytorch briefly...

**Comment by lucidrains** [comment]
@johahi oh bummer ok

i quite like your solution and think we should just make it official (with some minor changes). let me try to get this done in the next hour

i didn't know enformer had a successor! i'll have to read the borzoi paper when i get some time

**Comment by johahi** [comment]
Amazing, works now for me and hits the same numbers I reported up there for my fork, thanks

**Comment by lucidrains** [comment]
@johahi thank you for sharing your solution!
